### PR TITLE
A Tiny JS World pre-oop task

### PR DESCRIPTION
**Pull request by lalalilulu**
### Building a Tiny JS World

[Demo](https://lalalilulu.github.io/a-tiny-JS-world/) |
[Code base](https://github.com/lalalilulu/a-tiny-JS-world/tree/populate-world-l) 

The code is submitted in a dedicated feature branch.
Please, review.

### CODE REVIEW

File: submissions/lalalilulu/tiny-world-pre-oop/index.js [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+// Define your objects here
+
+class Inhabitant {
+
+    constructor(species, name, gender, saying, friends = []) {
+        this.species = species;
+        this.name = name;
+        this.gender = gender;
+        this.saying = saying;
+        this.friends = Array.from(friends).join(", ");
```
Comment (OleksiyRudenko) [inline submissions/lalalilulu/tiny-world-pre-oop/index.js:18]: What is the purpose of `.from` here?

File: submissions/lalalilulu/tiny-world-pre-oop/index.js [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+    introduce() {
+        return  super.introduce() + ` and I have ${this.legs} legs and ${this.hands} hands`;
+    }
+}
+
+class Animal extends Inhabitant{
+
+    constructor(species, name, gender, saying, friends, legs = 4) {
+        super(species, name, gender, saying, friends);
+        this.legs = legs;
```
Comment (OleksiyRudenko) [inline submissions/lalalilulu/tiny-world-pre-oop/index.js:43]: Looks like `legs` is a property shared by both humans and animals. This is a clear sign to generalize it into their base class.

File: submissions/lalalilulu/tiny-world-pre-oop/index.js [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+    }
+
+    introduce() {
```
Comment (OleksiyRudenko) [inline submissions/lalalilulu/tiny-world-pre-oop/index.js:46]: Not required to fix now, but you may consider overloading super base class `Object.toString`. This method is implictly called in a string context. Function `print` ensures string context for its arguments, thus `print(inhabitant)` is effectively equivalent to `print(inhabitant.toString())`

File: submissions/lalalilulu/tiny-world-pre-oop/index.js [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+        return  super.introduce() + ` and I have ${this.legs} legs`;
+    }
+}
+
+const dog = new Animal('dog', 'Toby', 'male', 'woof-woof!', ["Kate", "Greg"]);
+const cat = new Animal('cat', 'Mary', 'female', 'meow-meow!', [dog.name]);
+const man = new Human( 'Fred', 'male', 'Hi!', [dog.name, cat.name]);
+const woman = new Human('Katy', 'female', 'Hello!', [dog.name, cat.name, man.name]);
```
Comment (OleksiyRudenko) [inline submissions/lalalilulu/tiny-world-pre-oop/index.js:54]: Not required to fix within the pre-OOP. But will be required to fix later in post-OOP task.

Here is the problem:
Taxonomy doesn't look really consistent.
```
- Inhabitant
  |- Human
  \- Animal
```
Human represents a specific species, while Animal is not. Besides, humans are animals as well.

There are multiple ways to implement a hierarchy of classes representing domain taxonomy, but for consistency each level of hierarchy should reflect the same level of abstraction.

**Review by OleksiyRudenko — CHANGES_REQUESTED** [review]
@lalalilulu well done! Current implementation exceeds expectations at this stage.
You will be back to your Tiny JS World after a while with more strict requirements to meet.
Please answer questions below and fix where it looks appropriate to you.

**Comment by OleksiyRudenko** [comment]
> [Code base](https://github.com/lalalilulu/a-tiny-JS-world) branch: populate-world-l

FYI, you may use a direct URL to a branch.
E.g. `https://github.com/lalalilulu/a-tiny-JS-world/tree/populate-world-l`

File: submissions/lalalilulu/tiny-world-pre-oop/index.js (hunk 1 above)
Comment (lalalilulu) [reply submissions/lalalilulu/tiny-world-pre-oop/index.js]: Array.from() method was a possible way to handle a type error, if a single value is passed for the field ‘friends’ (not an object array).  

Uncaught TypeError: friends.join is not a function
    at new Inhabitant (index.js:19)
    at new Animal (index.js:30)
    at new Cat (index.js:57)
    at index.js:71

File: submissions/lalalilulu/tiny-world-pre-oop/index.js (hunk 2 above)
Comment (lalalilulu) [reply submissions/lalalilulu/tiny-world-pre-oop/index.js]: I didn't notice it... fix it

File: submissions/lalalilulu/tiny-world-pre-oop/index.js (hunk 4 above)
Comment (lalalilulu) [reply submissions/lalalilulu/tiny-world-pre-oop/index.js]: Thanks for the detailed explanation - I have changed classes hierarchy     
        Inhabitant
              | 
          Animal 
          /\     \
Human Dog  Cat

File: submissions/lalalilulu/tiny-world-pre-oop/index.js [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+        return`${this.saying} I am a ${this.species}, my name is ${this.name}, my gender is ${this.gender}. I have ${this.legs} legs. My friends: ${this.friends}`;
+    }
+}
+
+class Animal extends Inhabitant{
+
+    constructor(species, name, gender, saying, friends, legs = 4) {
+        super(species, name, gender, saying, friends, legs);
+    }
+}
```
Comment (OleksiyRudenko) [inline submissions/lalalilulu/tiny-world-pre-oop/index.js:32]: `Animal` class looks a bit redundant. Its only purpose is to provide a default value for legs count, which is anyway explicitly assigned by its child classes.
Not a real problem in fact.

File: submissions/lalalilulu/tiny-world-pre-oop/index.js [hunk 6]
```diff
@@ -0,0 +1,98 @@
...
+    constructor(name, gender, saying, friends) {
+        super('cat', name, gender, saying, friends, 4);
+
+    }
+}
+
+class CatWoman extends Animal{
+
+    constructor(name, saying, friends) {
+        super('cat', name, 'female', saying, friends, 4);
```
Comment (OleksiyRudenko) [inline submissions/lalalilulu/tiny-world-pre-oop/index.js:65]: It is all like a `Cat`: cat species, 4 legs, no hands.
Doesn't look like good reasons to have a dedicated class for this.
Good reasons would be if CatWoman was all like a human with the exception that it speaks like a cat.

File: submissions/lalalilulu/tiny-world-pre-oop/index.js [hunk 7]
```diff
@@ -0,0 +1,98 @@
...
+
+    }
+}
+
+const dog = new Dog('Toby', 'male', 'woof-woof!', ["Kate", "Greg"]);
+const cat = new Cat('Mary', 'female', 'meow-meow!', [dog.name]);
+const man = new Human( 'Fred', 'male', 'Hi!', [dog.name, cat.name]);
+const woman = new Human('Katy', 'female', 'Hello!', [dog.name, cat.name, man.name]);
+const catWoman = new CatWoman('Murka', 'meow-meow! Hello!', [cat.name, woman.name]);
```
Comment (OleksiyRudenko) [inline submissions/lalalilulu/tiny-world-pre-oop/index.js:74]: In real world friends are real inhabitants rather than just their names.

**Review by OleksiyRudenko — APPROVED** [review]
@lalalilulu well done!
Consider the comments below to improve your code when you're back to your tiny JS world (you will be after the JS OOP task).

**Pull request merged**